ql/src/java/org/apache/hadoop/hive/ql/parse/relnodegen/LateralViewPlan.java

```diff
@@ -0,0 +1,265 @@
+      if (child.getToken().getType() == HiveParser.TOK_TABALIAS ||
+          child.getToken().getType() == HiveParser.TOK_FUNCTION) {
+        continue;
+      }
+      String colAlias = BaseSemanticAnalyzer.unescapeIdentifier(child.getText().toLowerCase());
+      if (uniqueNames.contains(colAlias)) {
+        // Column aliases defined by query for lateral view output are duplicated
+        throw new SemanticException(ErrorMsg.COLUMN_ALIAS_ALREADY_EXISTS.getMsg(colAlias));
+      }
+      uniqueNames.add(colAlias);
+      colAliases.add(colAlias);
+    }
+
+    // if no column aliases were provided, just retrieve them from the return type
+    // of the udtf RexCall
+    if (colAliases.isEmpty()) {
+      colAliases.addAll(
+          Lists.transform(udtfCall.getType().getFieldList(), RelDataTypeField::getName));
+    }
+
+    // Verify that there is an alias for all the columns returned by the udtf call.
+    int udtfFieldCount = udtfCall.getType().getFieldCount();
+    if (colAliases.size() != udtfFieldCount) {
+      // Number of columns in the aliases does not match with number of columns
+      // generated by the lateral view
+      throw new SemanticException(ErrorMsg.UDTF_ALIAS_MISMATCH.getMsg(
+          "expected " + udtfFieldCount + " aliases " + "but got " + colAliases.size()));
+    }
+
+    return colAliases;
+  }
+
+  private List<RexNode> getOperandsForUDTF(ASTNode functionCall,
+      RowResolver inputRR) throws SemanticException {
+    List<RexNode> operands = new ArrayList<>();
+    TypeCheckCtx tcCtx = new TypeCheckCtx(inputRR, this.cluster.getRexBuilder(), false, false);
+    tcCtx.setUnparseTranslator(this.unparseTranslator);
+    // Start at 1 because value 0 is the function name.  Use the CalcitePlanner.genRexNode
+    // to retrieve the RexNode for all the function parameters.
+    for (int i = 1; i < functionCall.getChildren().size(); ++i) {
+      ASTNode functionParam = (ASTNode) functionCall.getChild(i);
+      operands.add(CalcitePlanner.genRexNode(functionParam, inputRR, tcCtx, this.conf));
+    }
+    return operands;
+  }
+
+  private RowResolver getOutputRR(RowResolver inputRR, RexCall udtfCall,
+      List<String> columnAliases, String lateralTableAlias) throws SemanticException {
+
+    RowResolver localOutputRR = new RowResolver();
+
+    // After calling RowResolver, outputRR will be mutated to contain the row resolver
+    // fields
+    if (!RowResolver.add(localOutputRR, inputRR)) {
+      LOG.warn("Duplicates detected when adding columns to RR: see previous message");
+    }
+
+    // The RexNode return value for a udtf is always a struct.
+    TypeInfo typeInfo = TypeConverter.convert(udtfCall.getType());
+    Preconditions.checkState(typeInfo instanceof StructTypeInfo);
+
+    StructTypeInfo typeInfos = (StructTypeInfo) typeInfo;
+    // Match up the column alias with the return value of the udtf and
+    // place in the outputRR
+    for (int i = 0, j = 0; i < columnAliases.size(); i++) {
```

Comment:
Should `j` be initialized before each do-while loop?

Reply:
That's a good question.  I was a little lazy here because I just grabbed the whole for loop from the original code and didn't want to rewrite this in case I missed something. But this piece of code does seem a bit awkward and maybe could use a little rewriting.

I think it's ok as it is though?  Essentially once it grabs the internal column name, that one will be "used".  If we reset "j" before the do while loop instead of the "for" loop, it will go back to zero but will proceed until it hits the "j+1" value anyway, I think, since all those internal columns will be in the outputRR.

Reply:
Ah ok, the value of `j` makes unique these internal column names in the RR. Let's keep this as it is.